native/python/src/fairseq2n/bindings/data/iterator_data_source.cc

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) Meta Platforms, Inc. and affiliates.
+// All rights reserved.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include "fairseq2n/bindings/data/iterator_data_source.h"
+
+namespace py = pybind11;
+
+namespace fairseq2n::detail {
+
+std::optional<data>
+iterator_data_source::next()
+{
+    py::gil_scoped_acquire acquire;
+    if (reloaded_) {
+        if (to_return_) {
+            reloaded_ = false;
+        }
+        return to_return_;
+    }
+    if (*iterator_ == py::iterator::sentinel()) {
+        return std::nullopt;
+    }
+    return (*iterator_)++->cast<py_object>();
+}
+
+void
+iterator_data_source::reset(bool)
+{
+    py::gil_scoped_acquire acquire;
+
+    reloaded_ = false;
+    reset_fn_(*iterator_);
```

Comment:
I think the whole `reset` approach to the Python iterator needs a bit rethinking. There are many cases where the user has no way to reset a Python iterator. Think for instance that they pass a `range(1234)` or `enumerate(other_iter)` to `read_iterator()`, there is no formal API in Python to "reset" such iterators. I think the safer approach would be for `reset_fn` (or whatever we call it) to return a new Python iterator. It will give the user the flexibility to return a new iterator (like in `range`, `enumerate`, or `itertools` iterators), or if they have a custom reset API, just call it, and return the same instance.